### PR TITLE
stat: replace custom reverse logic with slices.Reverse 

### PR DESCRIPTION
**Pull request by jbduncan**
This PR introduces a single modernization change from https://github.com/gonum/gonum/pull/1934. Benchmarks will follow.

Follow-on to https://github.com/gonum/gonum/pull/1934, which was my first attempt at this but was too large to properly benchmark.

Blocked by https://github.com/gonum/gonum/pull/1940.

### CODE REVIEW

**Comment by jbduncan** [comment]
Actually, `slices.ROC` has no benchmarks, so there's no need to run anything here.

**Comment by jbduncan** [comment]
As discussed in https://github.com/gonum/gonum/pull/1956#issuecomment-1988326786, I'll make some new benchmarks as soon as I can.

**Comment by jbduncan** [comment]
@kortschak Running the benchmarks in [82397bc](https://github.com/gonum/gonum/pull/1943/commits/82397bc9a111aafd53de61075b80ecfc2216f841), it looks like a few cases are better and worse than `master` by +-1-2%, with a slight increase in geomean by 0.48%.

Let me know if this is acceptable or not.

```
goos: darwin
goarch: arm64
pkg: gonum.org/v1/gonum/stat
                                                            │ benchmarks_stat_roc_old.txt │    benchmarks_stat_roc_new.txt     │
                                                            │           sec/op            │   sec/op     vs base               │
ROC/cutoffs=0,y=0,classes=0,weights=0-8                                       2.508n ± 0%   2.505n ± 0%       ~ (p=0.068 n=10)
ROC/cutoffs=0,y=10,classes=10,weights=0-8                                     124.2n ± 0%   123.9n ± 0%  -0.24% (p=0.043 n=10)
ROC/cutoffs=0,y=10,classes=10,weights=10-8                                    123.8n ± 0%   124.1n ± 0%  +0.24% (p=0.026 n=10)
ROC/cutoffs=0,y=1000,classes=1000,weights=0-8                                 8.991µ ± 0%   9.097µ ± 0%  +1.17% (p=0.000 n=10)
ROC/cutoffs=0,y=1000,classes=1000,weights=1000-8                              9.001µ ± 0%   9.076µ ± 1%  +0.83% (p=0.000 n=10)
ROC/cutoffs=0,y=100000,classes=100000,weights=0-8                             854.1µ ± 0%   856.0µ ± 0%       ~ (p=0.052 n=10)
ROC/cutoffs=0,y=100000,classes=100000,weights=100000-8                        851.1µ ± 0%   831.7µ ± 0%  -2.28% (p=0.000 n=10)
ROC/cutoffs=10,y=0,classes=0,weights=0-8                                      8.468n ± 0%   8.452n ± 0%  -0.18% (p=0.008 n=10)
ROC/cutoffs=10,y=10,classes=10,weights=0-8                                    118.7n ± 0%   120.2n ± 1%  +1.31% (p=0.000 n=10)
ROC/cutoffs=10,y=10,classes=10,weights=10-8                                   119.1n ± 0%   120.2n ± 0%  +0.92% (p=0.001 n=10)
ROC/cutoffs=10,y=1000,classes=1000,weights=0-8                                3.971µ ± 0%   3.972µ ± 0%       ~ (p=0.402 n=10)
ROC/cutoffs=10,y=1000,classes=1000,weights=1000-8                             3.980µ ± 0%   3.982µ ± 0%       ~ (p=0.400 n=10)
ROC/cutoffs=10,y=100000,classes=100000,weights=0-8                            477.4µ ± 1%   476.3µ ± 1%       ~ (p=0.393 n=10)
ROC/cutoffs=10,y=100000,classes=100000,weights=100000-8                       480.6µ ± 0%   477.6µ ± 0%  -0.62% (p=0.019 n=10)
ROC/cutoffs=1000,y=0,classes=0,weights=0-8                                    636.6n ± 1%   636.3n ± 0%       ~ (p=0.083 n=10)
ROC/cutoffs=1000,y=10,classes=10,weights=0-8                                  4.949µ ± 1%   5.049µ ± 1%  +2.03% (p=0.000 n=10)
ROC/cutoffs=1000,y=10,classes=10,weights=10-8                                 4.975µ ± 1%   5.051µ ± 1%  +1.53% (p=0.000 n=10)
ROC/cutoffs=1000,y=1000,classes=1000,weights=0-8                              9.604µ ± 0%   9.701µ ± 1%  +1.02% (p=0.000 n=10)
ROC/cutoffs=1000,y=1000,classes=1000,weights=1000-8                           9.625µ ± 1%   9.714µ ± 0%  +0.92% (p=0.003 n=10)
ROC/cutoffs=1000,y=100000,classes=100000,weights=0-8                          488.3µ ± 0%   494.0µ ± 0%  +1.17% (p=0.000 n=10)
ROC/cutoffs=1000,y=100000,classes=100000,weights=100000-8                     490.5µ ± 1%   495.2µ ± 0%  +0.96% (p=0.000 n=10)
ROC/cutoffs=100000,y=0,classes=0,weights=0-8                                  63.01µ ± 1%   62.79µ ± 0%  -0.36% (p=0.029 n=10)
ROC/cutoffs=100000,y=10,classes=10,weights=0-8                                467.0µ ± 4%   472.4µ ± 5%       ~ (p=0.190 n=10)
ROC/cutoffs=100000,y=10,classes=10,weights=10-8                               471.6µ ± 5%   472.8µ ± 3%       ~ (p=0.190 n=10)
ROC/cutoffs=100000,y=1000,classes=1000,weights=0-8                            502.7µ ± 1%   508.4µ ± 0%  +1.13% (p=0.002 n=10)
ROC/cutoffs=100000,y=1000,classes=1000,weights=1000-8                         503.1µ ± 3%   508.1µ ± 3%       ~ (p=0.218 n=10)
ROC/cutoffs=100000,y=100000,classes=100000,weights=0-8                        1.530m ± 0%   1.545m ± 1%  +0.97% (p=0.000 n=10)
ROC/cutoffs=100000,y=100000,classes=100000,weights=100000-8                   1.560m ± 0%   1.571m ± 0%  +0.70% (p=0.002 n=10)
geomean                                                                       15.66µ        15.74µ       +0.48%

```

**Comment by jbduncan** [comment]
Hi team, this PR is ready, so PTAL.

File: stat/bench_constants_test.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+// license that can be found in the LICENSE file.
+
+package stat
+
+const (
+	empty  = 0
+	small  = 10
+	medium = 1000
+	large  = 100000
+	huge   = 10000000
```
Comment (kortschak) [inline stat/bench_constants_test.go:12]: Please use e notation.

File: stat/roc_test.go [hunk 2]
```diff
@@ -324,3 +328,61 @@ func TestTOC(t *testing.T) {
...
+}
+
+func randomBools(l int, src rand.Source) []bool {
+	rnd := rand.New(src)
+	s := make([]bool, l)
+	for i := range s {
+		s[i] = rnd.Int31n(2) == 1
+	}
+	return s
+}
```
Comment (kortschak) [inline stat/roc_test.go:348]: Callees after callers.

File: stat/roc_test.go [hunk 3]
```diff
@@ -324,3 +328,61 @@ func TestTOC(t *testing.T) {
...
+			}
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				ROC(cutoffs, y, classes, weights)
+			}
+		})
+}
+
+func BenchmarkROC(b *testing.B) {
```
Comment (kortschak) [inline stat/roc_test.go:378]: Caller before callee.

**Review by kortschak — COMMENTED** [review]
Please move the benchmark addition prior to the logic change.

File: stat/bench_constants_test.go (hunk 1 above)
Comment (jbduncan) [reply stat/bench_constants_test.go]: Done!

File: stat/roc_test.go (hunk 2 above)
Comment (jbduncan) [reply stat/roc_test.go]: I believe that's done now!

File: stat/roc_test.go (hunk 3 above)
Comment (jbduncan) [reply stat/roc_test.go]: I believe this is done now!

**Comment by jbduncan** [comment]
@kortschak I believe this PR is ready whenever you are!

File: stat/bench_constants_test.go [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// Copyright ©2024 The Gonum Authors. All rights reserved.
```
Comment (kortschak) [inline stat/bench_constants_test.go:1]: Why did this move?

Comment (jbduncan) [reply stat/bench_constants_test.go:1]: I wanted to reuse these constants in `BenchmarkROC` in `stat/roc_test.go`, but leaving them in `stat/moments_bench_test.go` felt inappropriate, so I decided to move them into a common place.

I'm not sure of this project's conventions w.r.t. this, though, so I'm happy to move them back and duplicate them if you'd like.

Comment (kortschak) [reply stat/bench_constants_test.go:1]: Yeah, the move wasn't necessary, so let's revert that.

Comment (jbduncan) [reply stat/bench_constants_test.go:1]: Oh, sorry, I forgot to mention that I also introduced a new constant `empty`, which is only used in a different file, `stat/roc_test.go`. Any preferences on what I do with that constant? Leave it here, move it with the others to `stat/moments_bench_test.go`, or do something else entirely with it?

Comment (kortschak) [reply stat/bench_constants_test.go:1]: Let's just add the empty to the source location where the original constants were and revert the file addition.

Comment (jbduncan) [reply stat/bench_constants_test.go]: @kortschak Okay, done!

**Comment by jbduncan** [comment]
@kortschak I'm ready for a review once again.

**Review by kortschak — APPROVED** [review]

**Comment by jbduncan** [comment]
@kortschak Woo! Thank you for merging this in. 👍

**Pull request merged**